rfd/0127-encrypted-session-recordings.md

```diff
@@ -0,0 +1,849 @@
+not attempt to provision new keys or facilitate rotation.
+
+These configuration options are also accessible in the `teleport.yml` file
+configuration under the `auth_service` section:
+
+```yaml
+# teleport.yml
+auth_service:
+  session_recording: node
+  session_recording_config:
+    encryption:
+      enabled: true
+      manual_key_management:
+        enabled: true
+        active_keys:
+          - type: aws_kms
+            label: 'aws key ID'
+        rotated_keys: []
+```
+
+### Protobuf Changes
+
+```proto
+// api/proto/teleport/legacy/types/types.proto
+
+// EncryptionKeyPair is a keypair used for encrypting and decrypting data.
+message EncryptionKeyPair {
+  // public_key is the public encryption key.
+  bytes public_key = 1 [(gogoproto.jsontag) = "public_key"];
+  // PrivateKey is the private decryption key.
+  bytes private_key = 2 [(gogoproto.jsontag) = "private_key"];
+  // PrivateKeyType is the type of the private_key.
+  PrivateKeyType private_key_type = 3 [(gogoproto.jsontag) = "private_key_type"];
+  // Hash is the hash function to use during encryption/decryption operations.
+  // It maps directly to the possible values of crypto.Hash in the go crypto
+  // package.
+  uint32 hash = 4 [(gogoproto.jsontag) = "hash"];
+}
+
+// AgeEncryptionKey is a PEM encoded RSA4096 public key used for encrypting
+// session recordings.
+message AgeEncryptionKey {
+  bytes public_key = 1 [(gogoproto.jsontag) = "public_key"];
+}
+
+// SessionRecordingConfigStatusV2 contains the encryption_keys that should be
+// used during any recording encryption operation.
+message SessionRecordingConfigStatusV2 {
+  repeated AgeEncryptionKey encryption_keys  = 1 [
+    (gogoproto.jsontag) = "encryption_keys"
+  ];
+}
+
+
+// SessionRecordingEncryptionConfig configures if and how session recordings
+// should be encrypted.
+message SessionRecordingEncryptionConfig {
+  bool enabled = 1 [(gogoproto.jsontag) = "enabled"];
+}
+
+// SessionRecordingConfigSpecV2 is the actual data we care about
+// for SessionRecordingConfig.
+message SessionRecordingConfigSpecV2 {
+  // existing fields omitted
+
+  SessionRecordingEncryptionConfig encryption = 3 [
+    (gogoproto.jsontag) = "encryption"],
+    (gogoproto.nullable) = true
+  ];
+}
+
+// SessionRecordingConfigV2 contains session recording configuration.
+message SessionRecordingConfigV2 {
+  // existing fields omitted
+
+  // Status contains all of the current and rotated keys used for encrypted
+  // session recording
+  SessionRecordingConfigStatusV2 status = 6 [
+    (gogoproto.jsontag) = "status",
+    (gogoproto.nullable) = true
+  ];
+}
+```
+
+```proto
+// api/proto/teleport/recording_encryption/v1/recording_encryption.proto
+
+import "teleport/header/v1/metadata.proto";
+import "teleport/legacy/types/types.proto";
+
+// KeyState represents that possible states a RecordingEncryptionKey can be in.
+enum KeyState {
+  // Default KeyState
+  KEY_STATE_UNSPECIFIED = 0;
```

Comment:
Do not ascribe any meaning to the zero value of an enum, it's not a default, it's a missing value and if it's necessary to specify it and it's set to zero then the message should be treated as not valid (as if the value was set to some other value that's completely unknown).